### PR TITLE
Return lengths from feature and token collations

### PR DESCRIPTION
**Pull request by janvainer**
Summary:

* Return feature lengths from collate_features
* Add TokenCollater and corresponding test
* Add Standardize feature transform
* SpeechSynthesisDataset returns lengths for tokens and features
* SpeechSynthesisDataset applies feature transforms after collation
----

Problem: Where to pad features, normalize features, tokenize etc?

Proposal: Use stateful Collate functions that can be passed by user, but are pre-set with reasonable defaults.

Example:
```
collate_features = CollateFeatures(
    pad_mode="reflect",
    normalize_features=StandardizeFeatures(**cuts.compute_global_feature_stats())
)
dataset = SpeechSynthesisDataset(cuts, collate_features=collate_features)
```
Benefits:

1. Stateful collations allow user to pre-configure eg. normalization with pre-computed mean and std without using `partial`
2. More control over padding schemes, tokenization, normalization etc. 
3. Does not bloat already heavy CutSet with additional functionality
4. Allows user to create custom collations
5. Simple enough - can work with simple collation function without any state, can be completely be avoided by user if she is satisfied with the defaults.

More Ideas/Questions:
1. Should collation and online feature augmentation be treated on the same level? (eg they would be passed in a list of callables)I think maybe not, because collation returns also feature lengths
2. Should the normalization be considered online augmentation instead of part of Collation? (probably yes)

More generic suggestion:
```
collate_features = CollateFeatures(
    pad_mode="reflect",
)
transform_features = Transforms(
    StandardizeFeatures(**cuts.compute_global_feature_stats()),
    SpecAugment(some_other_args),
)

dataset = SpeechSynthesisDataset(cuts, collate_features=collate_features, transform_features=transform_features)
```

The tests may fail now, but I will fix that based on reactions to the proposal :)

### CODE REVIEW

File: lhotse/dataset/collation.py
Comment (danpovey) [inline lhotse/dataset/collation.py:44]: this doesn't work for me in torch 1.6.0 (this method of creating a tensor as torch.Tensor([list], dtype=int).
also int appears to map to torch.int64, better to use torch dtype.

File: lhotse/dataset/speech_synthesis.py
Comment (danpovey) [inline lhotse/dataset/speech_synthesis.py:45]: I don't understand how it manages to preserve the mapping between token-sequences and feature-sequences, with this unique_tokens() call.

Comment (janvainer) [reply lhotse/dataset/speech_synthesis.py:45]: The initialization creates a mapping from tokens to integers based on all used tokens in the dataset. This mapping is then used to map tokens loaded in the dataset to integers through the CollateTokens.__call__ method. Or maybe I do not understand the question correctly. Could you try to re-iterate?

File: lhotse/dataset/collation.py
Comment (janvainer) [reply lhotse/dataset/collation.py:44]: agreed. Btw I haven't tested all code aspects. This is more like a design idea - a draft PR.

Comment (janvainer) [reply lhotse/dataset/collation.py]: fixed

**Comment by pzelasko** [comment]
Thanks! I just skimmed through it - I will need to read more in-depth to develop a more formed opinion. I think this PR is helpful in converging on the design of these things.

I’m not sure if we really need to customize the collation fn - I think that any given task-specific dataset has a single „correct” method of collating the data? Then it makes sense to have it coded directly in dataset definition. OTOH one argument against that approach is that we might have single and multi channel versions of some tasks, and we will have to decide where to have separate dataset classes for that or a single one with input customization. I am leaning slightly towards the former — I had an attempt at a „generic” dataset class that is very customizable and I think it was confusing for others.

Last comment - another factor is also „on-the-fly” feature extraction, should there be a separate dataset class for that or another customization point for the existing classes. Again, I am slightly leaning towards having a lot of modularized functionality, so that writing each dataset classes takes 20~30 LOC — rather having 10+ arguments that customize a single class. I think it’s just harder for the end-users to get a grasp on that.

**Comment by janvainer** [comment]
_...any given task-specific dataset has a single „correct” method of collating the data_

I agree with this. But even if there is a single correct way how to do collations, there are small tweaks that need configuring. For example, the user may need to specify if the `<bos>` or the `<eos>` token should be used or what padding scheme should be used. I think the collation params could be passed directly to the dataset, and the augmentation features could be passed as a list of transforms. Continuing with speech synthesis example..

The args would be:
1. collations - only pass basic args such as `add_bos`, initialize inside dataset
2. online transforms - pass list of transforms objects

```
dataset =  SpeechSynthesisDataset(
    cuts,
    feature_transforms=Transforms(  # follows audiomentations-like api
        Rescale(0, 1),
    ),
    feature_pad_mode=reflect,
    use_bos=True,
    use_eos=True,
)

...
# the collation functions are configured internally inside the dataset without direct user control
feature_collate = CollateFeatures(pad=reflect, dtype=torch.float32), 
token_collate = CollateTokens(unique_tokens, use_bos=True, use_eos=False) 
```

If the above params could not be at least slightly customized, there would literaly be a separate dataset for every speech synthesis model (and they would be almost the same except for some small changes), which is not great either. How about a mix:

1. Have one very configurable SpeechSynthesisDataset like the original one that I posted with composable abstractions
2. Have wrapper datasets for specific speech synthesis tasks that would pre-configure the SpeechSynthesisDataset for the user with settings from the original papers.
    - FastSpeechDataset
    - SpeedySpeech dataset
    - Tacotron 2 dataset
    - ParallelWaveGan dataset etc.
It would work just like the dataset recipes that lhotse already provides - eg prepare_ljspeech. The recipes take specific dataset and return common representation - CutSet. The wrapper datasets would take specific speech synthesis subtask and return a preconfigutred common representation - the SpeechSynthesisDataset.

This would break eg in the case of FastSpeech or SpeedySpeech that also need **durations** of phonemes/chars for training which are not returned right now and also in multi-speaker case, where some **speaker embeddings** have to be used. But this could be solved with inheritance - FastSpeechDataset would call SpeechSynthesisDataset.__get__item__() and simply add its own additional keys to the batch dict.

File: lhotse/dataset/collation.py
Comment (pzelasko) [inline lhotse/dataset/collation.py:14]: style - either it's mimicking a function and then we should name it `collate_tokens`, or it's an object and then we should name it `TokenCollater` - I am leaning towards the latter

same for other classes of this type

File: lhotse/dataset/collation.py
Comment (pzelasko) [inline lhotse/dataset/collation.py:16]: I think this + post init is a hack in this case, because we don't need the original iterable after initialization. I think this should be a regular class with init that either accepts an iterable or a previously prepared mapping (or one of both)

File: lhotse/dataset/collation.py
Comment (pzelasko) [inline lhotse/dataset/collation.py:82]: Why do we need "reflect"?

File: lhotse/dataset/collation.py
Comment (pzelasko) [inline lhotse/dataset/collation.py:84]: as discussed this should be separated from this class

File: lhotse/dataset/collation.py
Comment (pzelasko) [inline lhotse/dataset/collation.py:207]: which style checker are you using?

**Review by pzelasko — COMMENTED** [review]
OK, I like where this is going. Some comments:

Re `SpeechSynthesisDataset` - I think this looks good, let's do this.

**Re phone durations:** - for now, Lhotse does not have an explicit representation for alignment, so getting this might be a bit difficult right now. We'll need to think how to represent it consistently with K2/Snowfall at some point (it will probably be an FSA, so it might be a bit tricky not to introduce a K2 dependency in Lhotse - @danpovey)

**Re speaker embeddings** - we can make it a convention that when a specific field is present in the SupervisionSegment object, we use it to read the speaker embeddings and automatically add them to batches. Similarly for other types of inputs and supervision, which might help us reduce the number of inherited dataset classes in total.


> But this could be solved with inheritance - FastSpeechDataset would call SpeechSynthesisDataset.__getitem__() and simply add its own additional keys to the returned batch dict.

I am not sure I like this - I think it makes the extension quite complicated since it's done after collation. But it's either some form of inheritance, or hooks with sth like "field" objects ([considered previously here](https://github.com/lhotse-speech/lhotse/pull/172/files#diff-6256dd88bea7f5803dffb53146cc885012a973dfa4d7b2b4824025bcae89a98bR1)), or "one-dataset-rules-them-all", or manual re-writing. TBH I think I still need to collect more data/use-cases to make a final decision on which way to go. So we can implement whichever of those but as far as I'm concerned it's still subject to change in the future.

File: lhotse/dataset/collation.py
Comment (janvainer) [reply lhotse/dataset/collation.py:16]: Fair point, I was a bit lazy with making regular init. Will change it to regular classes :)

File: lhotse/dataset/collation.py
Comment (janvainer) [reply lhotse/dataset/collation.py:14]: Agree, lets use `TokenCollater`

File: lhotse/dataset/collation.py
Comment (janvainer) [reply lhotse/dataset/collation.py:85]: Reflect can be used for batch norm to have unbiased estimate for batch variance (which usually happens if zero padding or other constant is used.) I put this here to show what configurations could be used and left the implementation for this blank. We can completely purge it for now.

File: lhotse/dataset/collation.py
Comment (janvainer) [reply lhotse/dataset/collation.py:210]: I used black with default settings.

Comment (pzelasko) [reply lhotse/dataset/collation.py:210]: I guess I'll need to make it the chosen autoformatter for Lhotse eventually, as the number of contributors grows. I wonder if it's possible to prevent it from changing unmodified code for now, as that could mess up git blame and makes PRs less readable in terms of actual changes.

Comment (janvainer) [reply lhotse/dataset/collation.py:210]: I will re-write the files anyway, so I ll keep the formatting as it was for this PR in the next push (probably tomorrow).

**Comment by janvainer** [comment]
Thanks for the review! 

**RE SpeechSynthesisDataset**: Great, I will polish the PR and resolve your style comments and implement the agreed API.
**RE Phone Durations**: I was thinking monotonic alignment (durations) could be stored as an array of integers and be loaded just like the features are loaded now, but not sure how this would interact with k2.
**RE Speaker embeddings**: I like the idea of putting them as fields in the SupervisionSegments. Autobatching of such fields sounds doable.
**RE Inheritance**: Good point with the after-collation. On the other hand, the collation happens separately for each field in the returned dict so inheritance would just chain one more collation on one more field or so. Could you let me know once the decision is made please? I would definitely be interested in pros and cons of the design in this case as its quite unclear which way would be the least painful.

**Comment by pzelasko** [comment]
Cool. Your have a good point about the collation too - I'm OK with inheritance for now if needed (maybe we can delay the need for that using custom fields in `SupervisionSegment` for now). You can store phone durations array in `SupervisionSegment.custom` field BTW, and make the dataset code look for it there.

**Comment by janvainer** [comment]
Ok it should be ready now. I decided not to overcomplicate things for now and left some of the things we mentioned for later.

File: lhotse/dataset/collation.py
Comment (pzelasko) [inline lhotse/dataset/collation.py:119]: Do you mind updating the rest of `collate_X` functions? There is an API mismatch at the moment, and they would have also benefitted from your changes.

File: lhotse/dataset/collation.py
Comment (pzelasko) [inline lhotse/dataset/collation.py:69]: Is that a fundamental limitation? I think we could just concatenate two utterances' texts.

File: lhotse/dataset/collation.py
Comment (pzelasko) [inline lhotse/dataset/collation.py:64]: it is a cosmetic change but consider sorting idx2token and then creating token2idx from that; it might be a bit faster to debug/lookup stuff

File: lhotse/dataset/diarization.py
Comment (pzelasko) [inline lhotse/dataset/diarization.py:64]: I think this dataset could possibly benefit from adding `features_lens` as a field in batch

File: lhotse/dataset/speech_synthesis.py
Comment (pzelasko) [inline lhotse/dataset/speech_synthesis.py:44]: I think it has to be at least `Sequence`, `Iterable` can be a one-pass iterator

File: lhotse/dataset/speech_synthesis.py
Comment (pzelasko) [inline lhotse/dataset/speech_synthesis.py:29]: mutable default arguments are dangerous; it is better to write `= None` and handle the case with `if x is None: self.x = []`

File: lhotse/dataset/collation.py
Comment (janvainer) [reply lhotse/dataset/collation.py:119]: Good point, I am off the computer already, so I will push an update tomorrow :)

File: lhotse/dataset/vad.py
Comment (pzelasko) [inline lhotse/dataset/vad.py:32]: can we add `features_lens` here?

File: lhotse/dataset/unsupervised.py
Comment (pzelasko) [inline lhotse/dataset/unsupervised.py:26]: can we add `features_lens` here?

File: test/dataset/test_collation.py
Comment (pzelasko) [inline test/dataset/test_collation.py:24]: would be good to assert some properties of tokens_batch and tokens_lens too (that they are int tensors, have the right shapes, and maybe specific values for lens)

File: test/dataset/test_speech_synthesis_dataset.py
Comment (pzelasko) [inline test/dataset/test_speech_synthesis_dataset.py:15]: Can we make it a parametrized test?

```python
@pytest.mark.parametrize('transform', [None, Standardize])
def test_speech_synthesis_dataset(cut_set, transform):
    transform = transform(cut_set) if transform is not None else None
    dataset = SpeechSynthesisDataset(cut_set, feature_transforms=transform)
```

**Review by pzelasko — COMMENTED** [review]
Looks good! I left some comments.

**Comment by pzelasko** [comment]
> Ok it should be ready now. I decided not to overcomplicate things for now and left some of the things we mentioned for later.

Sounds good to me - let's take it one step at a time :)

File: lhotse/dataset/collation.py
Comment (janvainer) [reply lhotse/dataset/collation.py:69]: Hm I guess we can. I can not think of a scenario when this would be a problem tbh.

File: lhotse/dataset/collation.py
Comment (janvainer) [reply lhotse/dataset/collation.py:64]: Thanks, good suggestion. Will adjust.

File: lhotse/dataset/diarization.py
Comment (janvainer) [reply lhotse/dataset/diarization.py:64]: I am thinking of adding features lens to most of the datasets, maybe except the K2 dataset. WYT?

File: lhotse/dataset/speech_synthesis.py
Comment (janvainer) [reply lhotse/dataset/speech_synthesis.py:44]: good catch, fixing

File: lhotse/dataset/speech_synthesis.py
Comment (janvainer) [reply lhotse/dataset/speech_synthesis.py:29]: Should it then be typed with Optional? `Optional[Sequence[Callable]] = None`?

File: lhotse/dataset/vad.py
Comment (janvainer) [reply lhotse/dataset/vad.py:32]: yes :)

File: lhotse/dataset/unsupervised.py
Comment (janvainer) [reply lhotse/dataset/unsupervised.py:26]: yes :D

File: test/dataset/test_collation.py
Comment (janvainer) [reply test/dataset/test_collation.py:24]: adding

File: test/dataset/test_speech_synthesis_dataset.py
Comment (janvainer) [reply test/dataset/test_speech_synthesis_dataset.py:15]: Good idea. Thanks!

File: lhotse/dataset/collation.py
Comment (janvainer) [reply lhotse/dataset/collation.py:64]: Looking more into it, sorting the idx2token would mean that the pad symbol would not have the index zero, which is default padding index eg in pytorch Embedding layer. Or am I missing something?

File: lhotse/dataset/speech_synthesis.py
Comment (pzelasko) [reply lhotse/dataset/speech_synthesis.py:29]: I think it's going to become difficult to read. I'm OK to omit Optional in such cases.

File: lhotse/dataset/collation.py
Comment (pzelasko) [reply lhotse/dataset/collation.py:64]: Oh sorry I didn't notice it is already sorted a few lines before. My bad!

**Comment by janvainer** [comment]
I pushed some changes.

- parameterized tests and more checks
- length is returned from more datasets

File: lhotse/dataset/speech_synthesis.py
Comment (pzelasko) [inline lhotse/dataset/speech_synthesis.py:49]: left-over print

File: test/dataset/test_speech_synthesis_dataset.py
Comment (pzelasko) [inline test/dataset/test_speech_synthesis_dataset.py:13]: actually we should add a third argument `[Standardize]` in there to ensure that it also works with sequence arguments

File: lhotse/dataset/speech_synthesis.py
Comment (janvainer) [reply lhotse/dataset/speech_synthesis.py:49]: thanks, fixing

File: test/dataset/test_speech_synthesis_dataset.py
Comment (janvainer) [reply test/dataset/test_speech_synthesis_dataset.py:13]: true, adding

**Comment by janvainer** [comment]
It seems like the tests in the CI sometimes fail even though I always test everything locally before pushing. Here are the cases I recall something weird happened:

1. All python versions except python3.6 passed. The 3.6 version passed due to some timeout error
2. Some randomized test failed, specifiaclly the `test/test_resample_randomized.py`

I am not able to reproduce any of these locally, but they happened in CI for some reason.

**Comment by pzelasko** [comment]
Sorry about that. The timeout error should already be fixed in latest master. The resample error is on my radar, I'll get to it soon. A fun fact is that the randomized tests failures can be replicated - the test output has a line like `use @reproduce_failure(...) decorator to reproduce`, you have to add that decorator on top of the function and you can easily replicate and debug then (but don't - I will take care of this particular bug).

**Comment by janvainer** [comment]
Cool, good to hear. I was not sure if this was being handled. :)

File: lhotse/dataset/collation.py
Comment (pzelasko) [reply lhotse/dataset/collation.py:115]: I think at the very least add lengths to collate audio below, and update TTS dataset to have `audio_lens` too. Multi-channel collations are ok the be left behind for now. 

I'll merge after that. Thanks!

Comment (janvainer) [reply lhotse/dataset/collation.py:115]: I added the audio_lens, adjusted test for the unsupervised audio dataset and added audio_lens to the speech synthesis dataset 👍🏻

**Review by pzelasko — APPROVED** [review]
Looks good to me, merging!

**Pull request merged**
